src/main/scala/is/hail/methods/MaximalIndependentSet.scala

```diff
@@ -0,0 +1,101 @@
+package is.hail.methods
+
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+import org.apache.spark.graphx._
+
+import scala.reflect.ClassTag
+
+object MaximalIndependentSet {
+
+  def apply[VD: ClassTag, ED: ClassTag](g: Graph[VD, ED]): Set[Long] = {
+    // Initially set each vertex to its own degree.
+    // Start a pregel run, everyone passing a (ID, degree) pair.
+    //  -On message reception, if current degree is greater than received degree, update status to to reflect this, alert everyone
+    //  -If current degree <= received degree, status does not change, don't bother sending alert.
+
+    //Every vertex needs to know
+    // -Whether it changed recently (some facet of old state must be stored)
+    // -What the highest degree message received has been
+    // -What vertex is in that highest degree message
+
+    var graph = Graph[((VertexId, Int), VertexId), ED](g.collectNeighborIds(EdgeDirection.Either)
+      .map{ case(v, neighbors) => (v, ((v, neighbors.size), -1L))}, g.edges: RDD[Edge[ED]])
```

Comment:
I think I know a rather efficient way to get the degree for *all* vertices:

```scala
private def updateVertexDegrees[VD: ClassTag, ED: ClassTag](g: Graph[VD, ED]): Graph[(Int, VertexId), ED] =
  Graph(g.vertices.leftZipJoin(g.degrees) { (v, _, degree) => (degree.getOrElse(0), v) }, g.edges)
```

The implementation of `leftZipJoin` only succeeds when both `VertexRDD`s have the same partitioning (`degrees` preserves the partitioning). Moreover, `leftZipJoin` eventually calls to an efficient looking `while` loop.